### PR TITLE
[IMPROVEMENT] Scroll performance improvement

### PR DESCRIPTION
**Pull request by Sameesunkaria**
@RocketChat/ios

This PR focuses on some performance improvements, mainly related to threading issues with Realm writes and JSON parsing on the main thread. The emoji replacement code for image emojis has also been rewritten (it should have a better performance characteristic, but i'm am not sure how it really performs against the old code).

This PR does not solve all the performance issues, i believe there are still performance enhancements that need to be done for handling the emoji code. Also, the `performBatchUpdates` call in `ChatViewController` is taking a significant amount of time.

These changes will help reduce most of the major hitches and (prolonged) hangs, but some still remain.

## Addendum:
I believe most of the performance related issues for the emoji replacement code are also fixed. What still needs some improvement is the code for drawing the cells, especially for the cells which have the indented text. (Such as replies)

I **really recommend a thorough review** of this PR, since it changes how the threading is being handled. Please try to test the build out on a real device as well, as some of the threading related issues are only incurred on a real device.

### CODE REVIEW

File: Rocket.Chat.xcodeproj/project.pbxproj [hunk 1]
```diff
@@ -2247,7 +2250,7 @@
 				TargetAttributes = {
 					41DF76DE1D2C50710028DBF8 = {
 						CreatedOnToolsVersion = 7.3.1;
-						DevelopmentTeam = S6UPZG7ZR3;
+						DevelopmentTeam = 8M5NP3VM5V;
```
Comment (artrmz) [inline Rocket.Chat.xcodeproj/project.pbxproj]: Need to revert this one

File: Rocket.Chat.xcodeproj/project.pbxproj [hunk 2]
```diff
@@ -3267,12 +3271,12 @@
...
 				CODE_SIGN_IDENTITY = "iPhone Developer";
 				"CODE_SIGN_IDENTITY[sdk=iphoneos*]" = "iPhone Developer";
-				DEVELOPMENT_TEAM = S6UPZG7ZR3;
+				DEVELOPMENT_TEAM = 8M5NP3VM5V;
 				INFOPLIST_FILE = Rocket.Chat/Info.plist;
 				IPHONEOS_DEPLOYMENT_TARGET = 10.0;
 				LD_RUNPATH_SEARCH_PATHS = "$(inherited) @executable_path/Frameworks";
 				OTHER_SWIFT_FLAGS = "$(inherited) -Onone -D DEBUG \"-D\" \"COCOAPODS\"";
-				PRODUCT_BUNDLE_IDENTIFIER = chat.rocket.ios;
+				PRODUCT_BUNDLE_IDENTIFIER = sam.chat.rocket.ios;
```
Comment (artrmz) [inline Rocket.Chat.xcodeproj/project.pbxproj]: Need to revert this one

File: Rocket.Chat.xcodeproj/project.pbxproj [hunk 3]
```diff
@@ -3290,11 +3294,11 @@
 				CODE_SIGN_ENTITLEMENTS = Rocket.Chat/Rocket.Chat.entitlements;
 				CODE_SIGN_IDENTITY = "iPhone Developer";
 				"CODE_SIGN_IDENTITY[sdk=iphoneos*]" = "iPhone Developer";
-				DEVELOPMENT_TEAM = S6UPZG7ZR3;
+				DEVELOPMENT_TEAM = 8M5NP3VM5V;
```
Comment (artrmz) [inline Rocket.Chat.xcodeproj/project.pbxproj]: Need to revert this one

File: Rocket.Chat.xcodeproj/project.pbxproj [hunk 4]
```diff
@@ -3290,11 +3294,11 @@
...
 				INFOPLIST_FILE = Rocket.Chat/Info.plist;
 				IPHONEOS_DEPLOYMENT_TARGET = 10.0;
 				LD_RUNPATH_SEARCH_PATHS = "$(inherited) @executable_path/Frameworks";
-				PRODUCT_BUNDLE_IDENTIFIER = chat.rocket.ios;
+				PRODUCT_BUNDLE_IDENTIFIER = sam.chat.rocket.ios;
```
Comment (artrmz) [inline Rocket.Chat.xcodeproj/project.pbxproj]: Need to revert this one

File: Rocket.Chat.xcodeproj/project.pbxproj [hunk 5]
```diff
@@ -3429,12 +3433,12 @@
 				CODE_SIGN_ENTITLEMENTS = Rocket.Chat/Rocket.Chat.entitlements;
 				CODE_SIGN_IDENTITY = "iPhone Developer";
 				"CODE_SIGN_IDENTITY[sdk=iphoneos*]" = "iPhone Developer";
-				DEVELOPMENT_TEAM = S6UPZG7ZR3;
+				DEVELOPMENT_TEAM = 8M5NP3VM5V;
```
Comment (artrmz) [inline Rocket.Chat.xcodeproj/project.pbxproj]: Need to revert this one

File: Rocket.Chat.xcodeproj/project.pbxproj [hunk 6]
```diff
@@ -3429,12 +3433,12 @@
...
 				INFOPLIST_FILE = Rocket.Chat/Info.plist;
 				IPHONEOS_DEPLOYMENT_TARGET = 10.0;
 				LD_RUNPATH_SEARCH_PATHS = "$(inherited) @executable_path/Frameworks";
 				OTHER_SWIFT_FLAGS = "$(inherited) \"-D\" \"COCOAPODS\" \"-D\" \"BETA\"";
-				PRODUCT_BUNDLE_IDENTIFIER = chat.rocket.ios;
+				PRODUCT_BUNDLE_IDENTIFIER = sam.chat.rocket.ios;
```
Comment (artrmz) [inline Rocket.Chat.xcodeproj/project.pbxproj]: Need to revert this one

**Comment by Sameesunkaria** [comment]
@artrmz Should be fixed.

File: Rocket.Chat/External/RCEmojiKit/NSAttributedString+CustomEmojis.swift [hunk 7]
```diff
@@ -41,13 +36,15 @@ extension NSAttributedString {
...
+                if let emoji = emojis[attributedString.attributedSubstring(from: range).string], let imageUrl = emoji.imageUrl {
+                    imageAttachment.contents = imageUrl.data(using: .utf8)
+                    let imageString = NSAttributedString(attachment: imageAttachment)
+                    attributedString.replaceCharacters(in: range, with: imageString)
+                }
             }
 
             return attributedString
-        }
+//        }
```
Comment (cardoso) [inline Rocket.Chat/External/RCEmojiKit/NSAttributedString+CustomEmojis.swift]: Can you remove this?

File: Rocket.Chat/Info.plist [hunk 8]
```diff
@@ -5,7 +5,7 @@
 	<key>CFBundleDevelopmentRegion</key>
 	<string>en</string>
 	<key>CFBundleDisplayName</key>
-	<string>Rocket.Chat+</string>
+	<string>Rocket.Test</string>
```
Comment (cardoso) [inline Rocket.Chat/Info.plist]: Undo this change.

**Comment by Sameesunkaria** [comment]
@cardoso Fixed. Also, did you take a look at the emoji replacement code?

**Comment by cardoso** [comment]
Yes @Sameesunkaria 

Looks like standard emojis are working, but custom emojis are not being replaced anymore. Can you take a look at that?

**Comment by Sameesunkaria** [comment]
Can you help me out with how to test for the custom emojis? like a sample replacement text..

**Comment by Sameesunkaria** [comment]
@cardoso The custom emoji should be fixed now. I have also added a test case for the same.

File: Rocket.Chat/External/RCEmojiKit/NSAttributedString+CustomEmojis.swift [hunk 9]
```diff
@@ -10,44 +10,42 @@ import Foundation
...
 
-            let regexPattern = ([emoji.shortname] + alternates).flatMap { $0.escapingRegex() }.reduce("") { $0.isEmpty ? ":\($1):" : "\($0)|:\($1):" }
-
-            guard let regex = try? NSRegularExpression(pattern: regexPattern, options: []) else { return attributedString }
-
-            let ranges = regex.matches(
-                in: attributedString.string,
-                options: [],
-                range: NSRange(location: 0, length: attributedString.length)
+        let ranges = regex.matches(
```
[the remark this answers is not in the document]
Comment (Sameesunkaria) [reply Rocket.Chat/External/RCEmojiKit/NSAttributedString+CustomEmojis.swift]: Not sure what's going on here, but hound reported no violations in the checks.

File: Rocket.Chat/Extensions/RealmExtension.swift [hunk 10]
```diff
@@ -26,6 +26,8 @@ extension Realm {
         }
     }
 
+    static let writeQueue = DispatchQueue(label: "chat.rocket.realm.write", qos: .background)
```
Comment (rafaelks) [inline Rocket.Chat/Extensions/RealmExtension.swift]: 👍

File: Rocket.Chat/Managers/Model/AuthSettingsManager.swift [hunk 11]
```diff
@@ -38,7 +38,8 @@ final class AuthSettingsManager {
                 return
             }
 
-            Realm.executeOnMainThread({ realm in
+            // Was executed on main
+            Realm.execute({ realm in
```
Comment (rafaelks) [inline Rocket.Chat/Managers/Model/AuthSettingsManager.swift]: You can remove the comment here. 👍

**Comment by rafaelks** [comment]
@Sameesunkaria Problem I found now, is that when you press the textField the scroll doesn't scroll with the keyboard when you're on the latest message. Can you change this behaviour? Only scroll to the bottom when focusing on keyboard if the user is at the bottom.
